rust/src/tx_builder.rs

```diff
@@ -265,118 +271,252 @@ impl TransactionBuilder {
-                    return Err(JsError::from_str("UTxO Balance Insufficient"));
-                }
             },
-            CoinSelectionStrategyCIP2::RandomImprove => {
-                fn add_random_input(
-                    s: &mut TransactionBuilder,
-                    rng: &mut rand::rngs::ThreadRng,
-                    available_inputs: &mut Vec<TransactionUnspentOutput>,
-                    input_total: &Value,
-                    output_total: &Value,
-                    added: &Value,
-                    needed: &Value
-                ) -> Result<(Value, Value, Value, Value, TransactionUnspentOutput), JsError> {
-                    let random_index = rng.gen_range(0..available_inputs.len());
-                    let input = available_inputs.swap_remove(random_index);
-                    // differing from CIP2, we include the needed fees in the targets instead of just output values
-                    let input_fee = s.fee_for_input(&input.output.address, &input.input, &input.output.amount)?;
-                    s.add_input(&input.output.address, &input.input, &input.output.amount);
-                    let new_input_total = input_total.checked_add(&input.output.amount)?;
-                    let new_output_total = output_total.checked_add(&Value::new(&input_fee))?;
-                    let new_needed = needed.checked_add(&Value::new(&input_fee))?;
-                    let new_added = added.checked_add(&input.output.amount)?;
-                    Ok((new_input_total, new_output_total, new_added, new_needed, input))
-                }
-                use rand::Rng;
-                if self.outputs.0.iter().any(|output| output.amount.multiasset.is_some()) {
-                    return Err(JsError::from_str("Multiasset values not supported by RandomImprove. Please use LargestFirst"));
-                }
-                // Phase 1: Random Selection
-                let mut associated_inputs: BTreeMap<TransactionOutput, Vec<TransactionUnspentOutput>> = BTreeMap::new();
-                let mut rng = rand::thread_rng();
-                let mut outputs = self.outputs.0.clone();
-                outputs.sort_by_key(|output| output.amount.coin);
-                for output in outputs.iter().rev() {
-                    let mut added = Value::new(&Coin::zero());
-                    let mut needed = output.amount.clone();
-                    while added < needed {
-                        if available_inputs.is_empty() {
-                            return Err(JsError::from_str("UTxO Balance Insufficient"));
+            CoinSelectionStrategyCIP2::LargestFirstMultiAsset => {
+                // indices into {available_inputs} for inputs that contain {policy_id}:{asset_name}
+                let mut available_indices = (0..available_inputs.len()).collect::<Vec<usize>>();
+                // run largest-fist by each asset type
+                if let Some(ma) = output_total.multiasset.clone() {
```

Comment:
I think clone might no be needed here since we only iterate over the data

Reply:
It was to placate the borrow checker since we pass in `output_total` by mutable borrow into the function which wouldn't be possible if we had an immutable borrow in that if above. It doesn't want there to be problems with iterator invalidation even if in this case we're only modifying the `coin` wrapped-u64 within `output_total`, but who knows what could happen if we happened to modify the `multiasset` parts within there with (de)allocatations.... I'm definitely not having any C++ flashbacks here so thanks borrow checker :)

If it becomes a performance issue we could pass the multi-asset part by immutable ref and the coin portion as mutable ref separately and then have to refactor the compare code used by Value to take in both by ref separately instead of as a whole Value, but that sounds like a lot of hassle and I don't think one clone for just the sum of outputs is going to make much of a difference at all.